pkg/deps/latest.go

```diff
@@ -0,0 +1,244 @@
+		if name != rootNodeName {
+			refs = append(refs, p.meta.ref())
+		}
+	}
+	return refs, nil
+}
+
+// graphBuilder constructs a possibly cyclic dependency graph.
+type graphBuilder struct {
+	picked            map[string]*node
+	componentVersions map[string]*sortedVersions
+	metaLookup        map[bundle.ComponentReference]*depMeta
+	searchOpts        *searchOpts
+}
+
+func shiftNode(ns []*node) (*node, []*node) {
+	return ns[0], ns[1:]
+}
+
+// visitNode constructs the latest graph for a single node. It returns the
+// nodes that need to be visited next or an error if things didn't work out.
+func (r *graphBuilder) visitNode(curNode *node) ([]*node, error) {
+	for success := false; !success; {
+		// We'll assume that adding the deps went swimmingly.
+		success = true
+
+		for _, dep := range curNode.meta.reqDeps {
+			picked, alreadyPicked := r.picked[dep.componentName]
+
+			var newChild *node
+			if alreadyPicked {
+				// We already picked a component with the same component name. Check to
+				// make sure the version works with this component's dependency
+				// requirements
+				successfulPick, err := r.processPicked(dep, curNode, picked)
+				if err != nil {
+					return nil, err
+				}
+				if !successfulPick {
+					success = false
+					break
+				}
+				newChild = picked
+			} else {
+				// Do selection for the first time. We'll pick the latest component
+				// version of the component that satisfies the requirements.
+				latest, err := r.pickLatest(dep, curNode)
+				if err != nil {
+					return nil, err
+				}
+				if latest == nil {
+					success = false
+					break
+				}
+				newChild = latest
+			}
+			if !newChild.meta.visibleTo(curNode.meta) {
+				return nil, fmt.Errorf("component %v is not visible to component %v", newChild.meta, curNode.meta)
+			}
+			curNode.children = append(curNode.children, newChild)
+		}
+
+		if !success {
+			err := r.tryDowngrade(curNode)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+	return curNode.children, nil
+}
+
+// processPicked checks to make sure an already picked component works with the
+// existing constraints
+func (r *graphBuilder) processPicked(dep *requestedDep, curNode, picked *node) (bool, error) {
+	if picked.meta.version.GTE(dep.version) {
+		// It works!
+		return true, nil
+	}
+	if dep.version.GT(picked.maxOfMinReq) {
+		picked.maxOfMinReq = dep.version
+	}
+	// It doesn't work. Indicate to the caller that a downgrade needs to be performed.
+	return false, nil
+}
+
+// pickLatest picks the latest component for a given dependency requirement.
+func (r *graphBuilder) pickLatest(dep *requestedDep, curNode *node) (*node, error) {
+	sorted, ok := r.componentVersions[dep.componentName]
+	if !ok {
+		return nil, fmt.Errorf("could not find sorted component versions for component name %q", dep.componentName)
+	}
+	latest, err := sorted.latest(r.searchOpts)
+	if err != nil {
+		return nil, err
+	}
+	if latest.version.LT(dep.version) {
```

Comment:
Move this control logic to the calling function - I'd prefer not to communicate state transitions via nil pointer